toolchain/lower/file_context.cpp

```diff
@@ -146,6 +160,240 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
   return std::move(llvm_module_);
 }
 
+auto FileContext::PairInserted(
+    SemIR::SpecificId specific_id1, SemIR::SpecificId specific_id2,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>& set_of_pairs)
+    -> bool {
+  if (specific_id1.index > specific_id2.index) {
+    std::swap(specific_id1.index, specific_id2.index);
+  }
+  std::pair<SemIR::SpecificId, SemIR::SpecificId> pair = {specific_id1,
+                                                          specific_id2};
+  auto insert_result = set_of_pairs.Insert(pair);
+  return insert_result.is_inserted();
+}
+
+auto FileContext::PairExists(
+    SemIR::SpecificId specific_id1, SemIR::SpecificId specific_id2,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>> set_of_pairs) -> bool {
+  if (specific_id1.index > specific_id2.index) {
+    std::swap(specific_id1.index, specific_id2.index);
+  }
+  std::pair<SemIR::SpecificId, SemIR::SpecificId> pair = {specific_id1,
+                                                          specific_id2};
+  return set_of_pairs.Contains(pair);
+}
+
+auto FileContext::CoalesceEquivalentSpecifics() -> void {
+  for (auto& specifics : lowered_specifics_) {
+    // i cannot be unsigned due to the comparison with a negative number when
+    // the specifics vector is empty.
+    for (int i = 0; i < static_cast<int>(specifics.size()) - 1; ++i) {
+      // This specific was already replaced, skip it.
```

Comment:
```suggestion
      // When this specific was already replaced, skip it.
```

Suggesting these "when" comments to reflect that it's describing the condition, rather than stating what's true when this comment was reached. This was particularly getting me on the third comment (line 226) where it switches from describing the condition ("When the two specifics are not equivalent") to instead describing the state ("The function type information matches")